File: go/internal/tools/linode_regions.go
package tools

import (
	"context"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
	"github.com/chadit/LinodeMCP/go/internal/linode"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
	"github.com/chadit/LinodeMCP/go/internal/toolschemas"
)

// NewLinodeRegionListTool creates a tool for listing Linode regions.
func NewLinodeRegionListTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newProtoListToolRawSchema(
		cfg,
		"linode_region_list",
		"Lists all available Linode regions (datacenters) with optional filtering by country or capabilities",
		"linode.mcp.v1.RegionListInput",
		func(ctx context.Context, client *linode.Client) ([]*linodev1.Region, error) {
			return client.ListRegionsProto(ctx)
		},
		[]listFilterParam[*linodev1.Region]{
			fieldFilter("country", "Filter regions by country code (e.g., 'us', 'de', 'jp')",
				func(r *linodev1.Region) string { return r.GetCountry() }),
			{
				paramName:   "capability",
				description: "Filter regions by capability (e.g., 'Linodes', 'Block Storage', 'GPU Linodes')",
				matchFunc:   filterRegionsByCapability,
			},
		},
		regionListResponse,
	)

	return tool, profiles.CapRead, handler
}

func regionListResponse(items []*linodev1.Region, count int32, filter *string) *linodev1.RegionListResponse {
	return &linodev1.RegionListResponse{Count: count, Filter: filter, Regions: items}
}

// NewLinodeRegionGetTool creates a tool for retrieving one Linode region.
func NewLinodeRegionGetTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool := mcp.NewToolWithRawSchema(
		"linode_region_get",
		"Gets one Linode region by region ID",
		toolschemas.Schema("linode.mcp.v1.RegionGetInput"),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		regionID, validationMessage := regionIDFromTool(&request)
		if validationMessage != "" {
			return mcp.NewToolResultError(validationMessage), nil
		}

		client, err := prepareClient(&request, cfg)
		if err != nil {
			return mcp.NewToolResultError(err.Error()), nil
		}

		region, getErr := client.GetRegionProto(ctx, regionID)
		if getErr != nil {
			return regionGetToolFailure(getErr)
		}

		return MarshalProtoToolResponse(region)
	}

	return tool, profiles.CapRead, handler
}

// NewLinodeRegionAvailabilityListTool creates a tool for listing compute type availability across regions.
func NewLinodeRegionAvailabilityListTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newProtoListToolRawSchema(
		cfg,
		"linode_region_availability_list",
		"Lists compute instance type availability across Linode regions",
		"linode.mcp.v1.RegionAvailabilityListInput",
		func(ctx context.Context, client *linode.Client) ([]*linodev1.RegionAvailability, error) {
			return client.ListRegionsAvailabilityProto(ctx)
		},
		nil,
		regionAvailabilityListResponse,
	)

	return tool, profiles.CapRead, handler
}

func regionAvailabilityListResponse(items []*linodev1.RegionAvailability, count int32, filter *string) *linodev1.RegionAvailabilityListResponse {
	return &linodev1.RegionAvailabilityListResponse{Count: count, Filter: filter, RegionAvailabilities: items}
}

// NewLinodeRegionAvailabilityGetTool creates a tool for listing compute type availability in one region.
func NewLinodeRegionAvailabilityGetTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool := mcp.NewToolWithRawSchema(
		"linode_region_availability_get",
		"Lists compute instance type availability for a Linode region",
		toolschemas.Schema("linode.mcp.v1.RegionAvailabilityGetInput"),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		regionID, validationMessage := regionAvailabilityRegionIDFromTool(&request)
		if validationMessage != "" {
			return mcp.NewToolResultError(validationMessage), nil
		}

		client, err := prepareClient(&request, cfg)
		if err != nil {
			return mcp.NewToolResultError(err.Error()), nil
		}

		availability, failureMessage := getRegionAvailability(ctx, client, regionID)
		if failureMessage != "" {
			return mcp.NewToolResultError(failureMessage), nil
		}

		return finishProtoList(&request, availability, nil, regionAvailabilityListResponse)
	}

	return tool, profiles.CapRead, handler
}

func regionGetToolFailure(failure error) (*mcp.CallToolResult, error) {
	return mcp.NewToolResultError("Failed to retrieve linode_region_get: " + failure.Error()), nil
}

func regionIDFromTool(request *mcp.CallToolRequest) (string, string) {
	regionID, validationMessage := requiredStringArg(request.GetArguments(), "region_id")
	if validationMessage != "" {
		return "", validationMessage
	}

	if err := validateRegionSlug(regionID); err != nil {
		return "", "region_id must be a lowercase region slug containing only letters, numbers, and hyphens"
	}

	return regionID, ""
}

func regionAvailabilityRegionIDFromTool(request *mcp.CallToolRequest) (string, string) {
	regionID, validationMessage := requiredStringArg(request.GetArguments(), "region_id")
	if validationMessage != "" {
		return "", validationMessage
	}

	if err := validateRegionSlug(regionID); err != nil {
		return "", "region_id " + err.Error()
	}

	return regionID, ""
}

func getRegionAvailability(ctx context.Context, client *linode.Client, regionID string) ([]*linodev1.RegionAvailability, string) {
	availability, err := client.GetRegionAvailabilityProto(ctx, regionID)
	if err != nil {
		return nil, "Failed to retrieve linode_region_availability_get: " + err.Error()
	}

	return availability, ""
}

func filterRegionsByCapability(regions []*linodev1.Region, capabilityFilter string) []*linodev1.Region {
	filtered := make([]*linodev1.Region, 0, len(regions))

	for i := range regions {
		for _, cap := range regions[i].GetCapabilities() {
			if strings.EqualFold(cap, capabilityFilter) {
				filtered = append(filtered, regions[i])

				break
			}
		}
	}

	return filtered
}

File: python/src/linodemcp/tools/linode_regions.py
"""Linode regions list tool."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from mcp.types import TextContent, Tool

from linodemcp.genpb.linode.mcp.v1 import region_pb2
from linodemcp.profiles import Capability
from linodemcp.tools.helpers import error_response, execute_tool
from linodemcp.tools.proto_response import (
    serialize_api_response,
    serialize_list_response,
)
from linodemcp.tools.toolschemas import schema

if TYPE_CHECKING:
    from linodemcp.linode import RetryableClient


def _is_region_id(value: str) -> bool:
    """Return True when value looks like a Linode region ID slug."""
    return bool(value) and all(c.isalnum() or c == "-" for c in value)


def create_linode_region_list_tool() -> tuple[Tool, Capability]:
    """Create the linode_region_list tool."""
    return Tool(
        name="linode_region_list",
        description=(
            "Lists all available Linode regions (datacenters) "
            "with optional filtering by country or capabilities"
        ),
        inputSchema={
            "type": "object",
            "properties": {
                "environment": {
                    "type": "string",
                    "description": (
                        "Linode environment to use (optional, defaults to 'default')"
                    ),
                },
                "country": {
                    "type": "string",
                    "description": "Filter regions by country code (e.g., 'us', 'de')",
                },
                "capability": {
                    "type": "string",
                    "description": (
                        "Filter regions by capability "
                        "(e.g., 'Linodes', 'Block Storage')"
                    ),
                },
            },
        },
    ), Capability.Read


async def handle_linode_region_list(
    arguments: dict[str, Any], cfg: Any
) -> list[TextContent]:
    """Handle linode_region_list tool request.

    Country is a case-insensitive exact match; capability is a case-insensitive
    exact match against the region's capabilities list, mirroring the Go list
    tool's fieldFilter/filterRegionsByCapability.
    """
    country_filter: str = arguments.get("country", "")
    capability_filter: str = arguments.get("capability", "")

    def _matches(region: dict[str, Any]) -> bool:
        country = str(region.get("country", ""))
        if country_filter and country.lower() != country_filter.lower():
            return False
        capabilities = region.get("capabilities", [])
        return not (
            capability_filter
            and not any(
                str(cap).lower() == capability_filter.lower() for cap in capabilities
            )
        )

    applied: list[str] = []
    if country_filter:
        applied.append(f"country={country_filter}")
    if capability_filter:
        applied.append(f"capability={capability_filter}")

    async def _call(client: RetryableClient) -> dict[str, Any]:
        raw = await client.get_raw("/regions")
        return serialize_list_response(
            raw,
            "regions",
            region_pb2.RegionListResponse(),
            filter_value=", ".join(applied) if applied else None,
            item_filter=_matches,
        )

    return await execute_tool(cfg, arguments, "retrieve Linode regions", _call)


def create_linode_region_get_tool() -> tuple[Tool, Capability]:
    """Create the linode_region_get tool."""
    return Tool(
        name="linode_region_get",
        description="Gets details for a specific Linode region",
        inputSchema=schema("linode.mcp.v1.RegionGetInput"),
    ), Capability.Read


async def handle_linode_region_get(
    arguments: dict[str, Any], cfg: Any
) -> list[TextContent]:
    """Handle linode_region_get tool request."""
    region_id = str(arguments.get("region_id", "")).strip()
    if not region_id:
        return error_response("region_id is required")
    if not _is_region_id(region_id):
        return error_response(
            "region_id must contain only letters, numbers, and hyphens"
        )

    async def _call(client: RetryableClient) -> dict[str, Any]:
        raw = await client.get_raw(f"/regions/{quote(region_id, safe='')}")
        return serialize_api_response(raw, region_pb2.Region())

    return await execute_tool(cfg, arguments, f"retrieve region {region_id}", _call)


def create_linode_region_availability_list_tool() -> tuple[Tool, Capability]:
    """Create the linode_region_availability_list tool."""
    return Tool(
        name="linode_region_availability_list",
        description="Lists compute instance type availability across Linode regions",
        inputSchema={
            "type": "object",
            "properties": {
                "environment": {
                    "type": "string",
                    "description": (
                        "Linode environment to use (optional, defaults to 'default')"
                    ),
                },
            },
        },
    ), Capability.Read


async def handle_linode_region_availability_list(
    arguments: dict[str, Any], cfg: Any
) -> list[TextContent]:
    """Handle linode_region_availability_list tool request."""

    async def _call(client: RetryableClient) -> dict[str, Any]:
        availability = await client.list_regions_availability()
        return serialize_list_response(
            {"data": availability},
            "region_availabilities",
            region_pb2.RegionAvailabilityListResponse(),
        )

    return await execute_tool(cfg, arguments, "retrieve regions availability", _call)


def create_linode_region_availability_get_tool() -> tuple[Tool, Capability]:
    """Create the linode_region_availability_get tool."""
    return Tool(
        name="linode_region_availability_get",
        description=(
            "Gets compute instance type availability for a specific Linode region"
        ),
        inputSchema={
            "type": "object",
            "properties": {
                "environment": {
                    "type": "string",
                    "description": (
                        "Linode environment to use (optional, defaults to 'default')"
                    ),
                },
                "region_id": {
                    "type": "string",
                    "description": "Region ID to check (for example, 'us-east')",
                },
            },
            "required": ["region_id"],
        },
    ), Capability.Read


async def handle_linode_region_availability_get(
    arguments: dict[str, Any], cfg: Any
) -> list[TextContent]:
    """Handle linode_region_availability_get tool request."""
    region_id = str(arguments.get("region_id", "")).strip()
    if not region_id:
        return error_response("region_id is required")
    if not _is_region_id(region_id):
        return error_response(
            "region_id must contain only letters, numbers, and hyphens"
        )

    async def _call(client: RetryableClient) -> dict[str, Any]:
        availability = await client.get_region_availability(region_id)
        return {
            "region_id": region_id,
            "count": len(availability),
            "availability": availability,
        }

    return await execute_tool(
        cfg, arguments, f"retrieve availability for region {region_id}", _call
    )
